### PR TITLE
helper.js: GetFolders -> GetModules, add package.json existence check

### PR DESCRIPTION
**Pull request by toxamiz**
Suggestion to https://github.com/govau/design-system-components/issues/696

Replace of `GetFolders` function.

### CODE REVIEW

File: scripts/helper.js [hunk 1]
```diff
@@ -145,17 +145,21 @@ const GetDepTree = ( name ) => {
...
  * @return {array}            - An array of names of each folder
  */
-const GetFolders = ( thisPath, verbose ) => {
+const GetModules = ( thisPath, verbose ) => {
 	try {
 		let folders = Fs.readdirSync( thisPath ).filter(
-				thisFile => Fs.statSync(`${ thisPath }/${ thisFile }`).isDirectory()
+				thisFile => {
+					let path = `${ thisPath }/${ thisFile }`;
+					return Fs.statSync(path).isDirectory() && Fs.existsSync(`${path}/package.json`)
```
Comment (adamzerella) [inline scripts/helper.js]: If we are going to refactor this function I think making it asynchronous would be really beneficial. Currently a lot of the code in `helper.js` is synchronous and as a good rule of thumb for any I/O this stuff should be async.

I like the readability changes but if we're updating this, we should make it _really_ nice

File: scripts/helper.js [hunk 2]
```diff
@@ -145,17 +145,21 @@ const GetDepTree = ( name ) => {
 
 
 /**
- * Get all folders within a given path
+ * Get all modules within a given path.
+ * Module is folder with package.json
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
 * The 'module' is a component folder
```

File: scripts/helper.js [hunk 3]
```diff
@@ -145,17 +145,21 @@ const GetDepTree = ( name ) => {
 
 
 /**
- * Get all folders within a given path
+ * Get all modules within a given path.
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
 * Get all modules within a given path
```

None of the other function descriptions have a full-stop, to be consistent we should leave it out

File: scripts/helper.js [hunk 4]
```diff
@@ -145,17 +145,21 @@ const GetDepTree = ( name ) => {
...
  *
  * @return {array}            - An array of names of each folder
  */
-const GetFolders = ( thisPath, verbose ) => {
+const GetModules = ( thisPath, verbose ) => {
 	try {
 		let folders = Fs.readdirSync( thisPath ).filter(
-				thisFile => Fs.statSync(`${ thisPath }/${ thisFile }`).isDirectory()
+				thisFile => {
+					let path = `${ thisPath }/${ thisFile }`;
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
					let path = Path.normalize( `${ thisPath }/${ thisFile }` );
```

We've had issues in the past with pathing across platforms, normalize should format this for us

File: scripts/helper.js (hunk 1 above)
Comment (toxamiz) [reply scripts/helper.js]: Yes, there are a lot of  stuff which can be run asynchronous.

OK, i can "promisify" this function and change calling code.

**Comment by toxamiz** [comment]
updated to async version

File: scripts/helper.js [hunk 5]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
-		);
-
-		return ['core', ...folders ]; // moving core to top
-	}
-	catch( error ) {
-		return [];
-	}
+const GetModules = ( thisPath, verbose ) => {
+	return new Promise(resolve => {
+		Readdir(thisPath)
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
		Readdir( thisPath )
```

File: scripts/helper.js [hunk 6]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
-				thisFile => thisFile !== 'core'
-		);
-
-		return ['core', ...folders ]; // moving core to top
-	}
-	catch( error ) {
-		return [];
-	}
+const GetModules = ( thisPath, verbose ) => {
+	return new Promise(resolve => {
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
	return new Promise( resolve => {
```

File: scripts/helper.js [hunk 7]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
+		Readdir(thisPath)
+			.then(dirContent => {
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
			.then( dirContent => {
```

File: scripts/helper.js [hunk 8]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
+				let folders = dirContent
+					.filter(folder => {
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
					.filter( folder => {
```

File: scripts/helper.js [hunk 9]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
+						let path = Path.normalize( `${ thisPath }/${ folder }` );
+						return Fs.statSync( path ).isDirectory() && Fs.existsSync( Path.normalize( `${path}/package.json` ) )
```
Comment (adamzerella) [inline scripts/helper.js]: This function could be refactored to read nicer and is also still synchronous and IO based.

File: scripts/helper.js [hunk 10]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
+					})
+					.filter(folder => folder !== 'core');
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
					.filter( folder => folder !== 'core' );
```

File: scripts/helper.js [hunk 11]
```diff
@@ -145,26 +147,29 @@ const GetDepTree = ( name ) => {
...
+
+				resolve(['core', ...folders ]); // moving core to top
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
				resolve( [ 'core', ...folders ] ); // moving core to top
```

File: scripts/helper.js [hunk 12]
```diff
@@ -155,21 +156,27 @@ const GetDepTree = ( name ) => {
...
-	})
+const GetModules = async ( thisPath, verbose ) => {
+	try {
+		let dirContent = await Readdir( thisPath );
+		let folders = ( await Promise.all(
+			dirContent.map( dirItem => {
+				let pkg = Path.normalize( `${ thisPath }/${ dirItem }/package.json` );
+				return ( async () => {
+					try {
+						return await Access( pkg ) || dirItem;
```
Comment (toxamiz) [inline scripts/helper.js]: This is not obvious behavior.

**Comment by toxamiz** [comment]
Hey, @adamzerella 

Updated to async/await syntax. But there is one big question, see below.

> Since we are refactoring we should also update the Fs.statSync and Fs.existsSync functions.

Async `Fs.exists` is deprecated. The only way to check file existence asynchronously is to use `Fs.access`. The main issue is `Fs.access` provides not obvious behavior, because it's a void function (return undefined), that throws an error if file is not exists. And we don't need to check that `package/component` is a directory via `Fs.stat` and straight check access to `package/component/package.json`.

What you think?

> Rather than promisifying the functions, we'd prefer, for readability sake to use the async/await syntax. Node has an experimental API at the moment that wraps the fs module in promises

It is node 10 experimental API, but your minimum requirement is node 8 ([readme](https://github.com/govau/design-system-components/blob/master/README.md#how-to-use)).

File: scripts/helper.js [hunk 13]
```diff
@@ -145,24 +148,33 @@ const GetDepTree = ( name ) => {
...
+				} )()
+			} )
+		) )
+		.filter( folder => folder !== '' && folder !== 'core' );
 
-		return ['core', ...folders ]; // moving core to top
-	}
-	catch( error ) {
+		return [ 'core', ...folders ]; // moving core to top
+	} catch (e) {
```
Comment (adamzerella) [inline scripts/helper.js]:
```suggestion
	} catch ( error ) {
	console.error( error );
```

Comment (adamzerella) [reply scripts/helper.js]: It'd be nice if we log this error just in case

**Comment by toxamiz** [comment]
Updated.

 >  I'm curious if the GetModules() function needs to be updated now that it is async and other functions that call it like init: () now being init: async ()

Not necessary for now. All init functions are used synchronously

**Comment by adamzerella** [comment]
@toxamiz I'm going to close this PR as it doesn't resolve the original issue it was referencing #696. It still has some outstanding formatting issues, this isn't your fault. As you brought up in #710 we should have linting functionality to address this but that carries over other issues as we've previously discussed. 

The `helper.js` file is going to be refactored in the future and possibly removed, we just aren't there yet.

**Pull request closed without merging**